vm/src/protocol/mapping.rs

```diff
@@ -1,126 +1,167 @@
+    common::lock::OnceCell,
+    function::IntoPyResult,
+    IdProtocol, PyObject, PyObjectPayload, PyObjectRef, PyObjectView, PyResult, TypeProtocol,
     VirtualMachine,
 };
-use std::borrow::Borrow;
 
 // Mapping protocol
 // https://docs.python.org/3/c-api/mapping.html
 #[allow(clippy::type_complexity)]
-#[derive(Default)]
+#[derive(Default, Copy, Clone)]
 pub struct PyMappingMethods {
-    pub length: Option<fn(PyObjectRef, &VirtualMachine) -> PyResult<usize>>,
-    pub subscript: Option<fn(PyObjectRef, PyObjectRef, &VirtualMachine) -> PyResult>,
+    pub length: Option<fn(&PyMapping, &VirtualMachine) -> PyResult<usize>>,
+    pub subscript: Option<fn(&PyMapping, &PyObject, &VirtualMachine) -> PyResult>,
     pub ass_subscript:
-        Option<fn(PyObjectRef, PyObjectRef, Option<PyObjectRef>, &VirtualMachine) -> PyResult<()>>,
+        Option<fn(&PyMapping, &PyObject, Option<PyObjectRef>, &VirtualMachine) -> PyResult<()>>,
 }
 
-#[derive(Debug, Clone)]
-#[repr(transparent)]
-pub struct PyMapping<T = PyObjectRef>(T)
-where
-    T: Borrow<PyObject>;
-
-impl PyMapping<PyObjectRef> {
-    pub fn check(obj: &PyObject, vm: &VirtualMachine) -> bool {
-        obj.class()
-            .mro_find_map(|x| x.slots.as_mapping.load())
-            .map(|f| f(obj, vm).subscript.is_some())
-            .unwrap_or(false)
+#[derive(Clone)]
+pub struct PyMapping<'a> {
+    pub obj: &'a PyObject,
+    methods: OnceCell<PyMappingMethods>,
+}
+
+impl<'a> From<&'a PyObject> for PyMapping<'a> {
+    fn from(obj: &'a PyObject) -> Self {
+        Self {
+            obj,
+            methods: OnceCell::new(),
+        }
     }
+}
 
-    pub fn methods(&self, vm: &VirtualMachine) -> PyMappingMethods {
-        let obj_cls = self.0.class();
-        for cls in obj_cls.iter_mro() {
-            if let Some(f) = cls.slots.as_mapping.load() {
-                return f(&self.0, vm);
-            }
+impl AsRef<PyObject> for PyMapping<'_> {
+    fn as_ref(&self) -> &PyObject {
+        self.obj
+    }
+}
+
+impl<'a> PyMapping<'a> {
+    pub fn with_methods(obj: &'a PyObject, methods: PyMappingMethods) -> Self {
+        Self {
+            obj,
+            methods: OnceCell::from(methods),
+        }
+    }
+
+    pub fn try_protocol(obj: &'a PyObject, vm: &VirtualMachine) -> PyResult<Self> {
+        let zelf = Self::from(obj);
+        if zelf.has_protocol(vm) {
+            Ok(zelf)
+        } else {
+            Err(vm.new_type_error(format!("{} is not a mapping object", zelf.obj.class())))
         }
-        PyMappingMethods::default()
     }
 }
 
-impl<T> PyMapping<T>
-where
-    T: Borrow<PyObject>,
-{
-    pub fn new(obj: T) -> Self {
-        Self(obj)
+impl PyMapping<'_> {
+    // PyMapping::Check
+    pub fn has_protocol(&self, vm: &VirtualMachine) -> bool {
+        self.methods(vm).subscript.is_some()
+    }
+
+    pub fn obj_as<T: PyObjectPayload>(&self) -> &PyObjectView<T> {
```

Comment:
This helper function is used to downcast the object to the type who provide the methods and should not used by any case if the type is not certain. maybe add the comment will be better than mark it unsafe because it is used in each methods to get the instance of Self?

Reply:
That sounds like this method is strictly required to be call like `obj_as::<Self>` always. In that case, the function is unsafe beacuse the guarantee of `Self` came from user, not from automatic constraints.
If you want to have a safe version of the function, I think a trait can wrap it safely. Maybe `AsMapping` can be the place.

```rust
trait AsMapping {
    fn mapping_as(mapping: ...) -> &PyObjectView<Self> {
        unsafe { mapping.obj_as::<Self>() }
    }
   ...
}
```